### PR TITLE
Delegation

### PR DESCRIPTION
**Pull request by eb-concordium**
## Purpose

To support new transactions (together with their costs) and special transaction outcomes introduced in Protocol version 4. 

## Changes

- The transaction cost endpoint has been updated to cover the all possible costs of the newly introduced `configure baker` and `configure delegation` transactions. 
- Existing endpoint `v0/accBalance` have been updated to additionally show if the account is a baker or delegator. 
- New endpoints `v0/bakerPool/n`, `v0/chainParameters` and `v0/nextPayday` have been added.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,6 +22,8 @@ The wallet proxy provides the following endpoints:
 * `GET /v0/global`: get the cryptographic parameters obtained from the node it is connected to
 * `GET /v0/ip_info`: get the identity providers information, including links for
   submitting initial identity issuance requests.
+* `GET /v0/bakerPool/{bakerId}`: get the status of a baker pool given the baker ID.
+* `GET /v0/poolParameters`: get the pool parameters.
```
Comment (abizjak) [inline README.md:26]: This is no longer correct.

File: src/Proxy.hs [hunk 2]
```diff
@@ -1135,3 +1295,39 @@ getHealthR =
...
+      $(logInfo) "Successfully got baker pool status."
+      sendResponse v
+  where
+    doGetBaker = withLastFinalBlockHash Nothing (getPoolStatus (BakerId $ AccountIndex bid) False)
+
+getChainParametersR :: Handler TypedContent
+getChainParametersR =
+    runGRPC doGetParameters $ \(v :: Parameters.ChainParameters' 'ChainParametersV1) -> do
+      $(logInfo) "Successfully got chain parameters."
+      sendResponse $ toJSON v
```
Comment (abizjak) [inline src/Proxy.hs:1312]: It would be useful to have a better error message here in case the node is running the old protocol version at the moment.

**Comment by shjortConcordium** [comment]
A configureDelegation transaction that is rejected due PoolClosed (trying to delegate to a closed pool), causes an error, when the account's transactions are queried.
Currently on stagenet, this query:
`https://wallet-proxy.stagenet.concordium.com/v1/accTransactions/3BpVX13dw29JruyMzCfde96hoB7DtQ53WMGVDMrmPtuYAbzADj`

Results in this:

> 11/Apr/2022:07:58:35 +0000 [Error] Error decoding transaction: parsing Concordium.Types.Execution.RejectReason failed, expected tag field to be one of ["ModuleNotWF","ModuleHashAlreadyExists","InvalidAccountReference","InvalidInitMethod","InvalidReceiveMethod","InvalidModuleReference","InvalidContractAddress","RuntimeFailure","AmountTooLarge","SerializationFailure","OutOfEnergy","RejectedInit","RejectedReceive","NonExistentRewardAccount","InvalidProof","AlreadyABaker","NotABaker","InsufficientBalanceForBakerStake","StakeUnderMinimumThresholdForBaking","BakerInCooldown","DuplicateAggregationKey","NonExistentCredentialID","KeyIndexAlreadyInUse","InvalidAccountThreshold","InvalidCredentialKeySignThreshold","InvalidEncryptedAmountTransferProof","InvalidTransferToPublicProof","EncryptedAmountSelfTransfer","InvalidIndexOnEncryptedTransfer","ZeroScheduledAmount","NonIncreasingSchedule","FirstScheduledReleaseExpired","ScheduledSelfTransfer","InvalidCredentials","DuplicateCredIDs","NonExistentCredIDs","RemoveFirstCredential","CredentialHolderDidNotSign","NotAllowedMultipleCredentials","NotAllowedToReceiveEncrypted","NotAllowedToHandleEncrypted","MissingBakerAddParameters","UnexpectedBakerRemoveParameters","CommissionsNotInRangeForBaking","AlreadyADelegator","InsufficientBalanceForDelegationStake","MissingDelegationAddParameters","UnexpectedDelegationRemoveParameters","DelegatorInCooldown","NotADelegator","DelegationTargetNotABaker","StakeOverMaximumThresholdForPool","PoolWouldBecomeOverDelegated"], but found tag "PoolClosed" @(wallet-proxy-0.13.0-73ejI6ZfnRe9vNLXyILxHY:Proxy src/Proxy.hs:796:17)

**Comment by orhoj** [comment]
> 

Make sure to add all the new error codes as `InsufficientDelegationStake` also results in an error.

File: src/Proxy.hs [hunk 3]
```diff
@@ -52,9 +52,11 @@ import Data.Time.Clock as Clock
 import Paths_wallet_proxy (version)
 import Concordium.Types
 import Concordium.Types.Queries
+import Concordium.Types.Accounts
 import Concordium.Types.HashableTo
 import Concordium.Types.Transactions
 import Concordium.Types.Execution
+import qualified Concordium.Types.Parameters as Parameters
```
Comment (abizjak) [inline src/Proxy.hs:59]: Unused import.

File: README.md [hunk 4]
```diff
@@ -22,6 +22,8 @@ The wallet proxy provides the following endpoints:
 * `GET /v0/global`: get the cryptographic parameters obtained from the node it is connected to
 * `GET /v0/ip_info`: get the identity providers information, including links for
   submitting initial identity issuance requests.
+* `GET /v0/bakerPool/{bakerId}`: get the status of a baker pool given the baker ID.
+* `GET /v0/chainParameters`: get the chain parameters.
```
Comment (abizjak) [inline README.md:26]: `/v0/nextPayday` is missing from this list.

**Review by abizjak — DISMISSED** [review]

**Review by abizjak — APPROVED** [review]

**Pull request merged**